src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs

```diff
@@ -0,0 +1,375 @@
+
+        // In the order we wish to detect a derived type.
+        private static readonly Type[] s_genericInterfacesWithAddMethods = new Type[]
+        {
+            typeof(IDictionary<,>),
+            typeof(ICollection<>),
+        };
+
+        // In the order we wish to detect a derived type.
+        private static readonly Type[] s_nonGenericInterfacesWithAddMethods = new Type[]
+        {
+            typeof(IDictionary),
+            typeof(IList),
+        };
+
+        // In the order we wish to detect a derived type.
+        private static readonly Type[] s_genericInterfacesWithoutAddMethods = new Type[]
+        {
+            typeof(IReadOnlyDictionary<,>),
+            typeof(IReadOnlyCollection<>),
+            typeof(IReadOnlyList<>),
+            typeof(IEnumerable<>),
+        };
+
+        public static Type GetImplementedCollectionType(Type queryType)
+        {
+            Debug.Assert(queryType != null);
+
+            if (queryType.IsAbstract ||
+                queryType.IsInterface ||
+                !(typeof(IEnumerable).IsAssignableFrom(queryType)) ||
+                queryType == typeof(string) ||
+                queryType.IsArray ||
+                queryType == typeof(object) ||
+                IsNativelySupportedCollection(queryType))
+            {
+                return queryType;
+            }
+
+            Type baseType = queryType.GetTypeInfo().BaseType;
+
+            // Check if the base type is a supported concrete collection.
+            if (IsNativelySupportedCollection(baseType))
+            {
+                return baseType;
+            }
+
+            // Try generic interfaces with add methods.
+            foreach (Type candidate in s_genericInterfacesWithAddMethods)
+            {
+                Type derivedGeneric = ExtractGenericInterface(queryType, candidate);
+                if (derivedGeneric != null)
+                {
+                    return derivedGeneric;
+                }
+            }
+
+            // Try non-generic interfaces with add methods.
+            foreach (Type candidate in s_nonGenericInterfacesWithAddMethods)
+            {
+                if (candidate.IsAssignableFrom(queryType))
+                {
+                    return candidate;
+                }
+            }
+
+            // Try generic interfaces without add methods
+            foreach (Type candidate in s_genericInterfacesWithoutAddMethods)
+            {
+                Type derivedGeneric = ExtractGenericInterface(queryType, candidate);
+                if (derivedGeneric != null)
+                {
+                    return derivedGeneric;
+                }
+            }
+
+            // Try non-generic interfaces without add methods
+            if (typeof(IEnumerable).IsAssignableFrom(queryType))
+            {
+                return typeof(IEnumerable);
+            }
+            else if (typeof(ICollection).IsAssignableFrom(queryType))
```

Comment:
nit: don't need `else` here (or add `else` below to make it consistent)

Reply:
I think we do because we want to return the `queryType` below this if it doesn't implement `ICollection`.

Also, consistent with what?